android/src/main/java/com/liferay/mobile/android/auth/refresh/CookieAuthenticationRefreshHandler.java

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ * <p>
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ * <p>
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.mobile.android.auth.refresh;
+
+import com.liferay.mobile.android.auth.Authentication;
+import com.liferay.mobile.android.auth.CookieSignIn;
+import com.liferay.mobile.android.auth.CookieSignIn.CookieCallback;
+import com.liferay.mobile.android.auth.SessionCallback;
+import com.liferay.mobile.android.auth.basic.CookieAuthentication;
+import com.liferay.mobile.android.service.Session;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class CookieAuthenticationRefreshHandler implements
+        AuthenticationRefreshHandler {
+
+    @Override
+    public Session refreshAuthentication(final Session session,
+        final SessionCallback callback) throws Exception {
+
+        if (shouldCheckCookieExpiration(session.getAuthentication())) {
+            CookieAuthentication authentication =
+                    (CookieAuthentication) session.getAuthentication();
+
+            lock.lock();
+
+            if (shouldRefreshCookie(
+                    authentication.getLastCookieRefresh(),
+                    authentication.getCookieExpirationTime())) {
+
+                if (callback == null) {
+                    try {
+                        return CookieSignIn.signIn(session);
+                    } finally {
+                        lock.unlock();
+                    }
+                } else {
+                    CookieSignIn.signIn(session, new CookieCallback() {
```

Comment:
Can this fail with exception, like above?

Reply:
In this case if there is an error, it will be notified to the callback

Reply:
Yeah, I mean if it can fail outside the callback. If something fails because NPE/validation before the process controlled by the callback